File: projectforge-business/src/main/java/org/projectforge/sms/SMSSender.java
package org.projectforge.sms;

import org.apache.commons.collections.MapUtils;
import org.apache.commons.httpclient.HttpClient;
import org.apache.commons.httpclient.HttpException;
import org.apache.commons.httpclient.HttpMethodBase;
import org.apache.commons.httpclient.NameValuePair;
import org.apache.commons.httpclient.methods.GetMethod;
import org.apache.commons.httpclient.methods.PostMethod;
import org.apache.commons.lang3.StringUtils;
import org.projectforge.common.StringHelper;
import org.springframework.beans.factory.annotation.Value;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.Map;

public class SMSSender {
  private static transient final org.slf4j.Logger log = org.slf4j.LoggerFactory.getLogger(SMSSender.class);

  public enum HttpResponseCode {SUCCESS, NUMBER_ERROR, MESSAGE_TO_LARGE, MESSAGE_ERROR, UNKNOWN_ERROR}

  private SMSSenderConfig config;

  public SMSSender(SMSSenderConfig config) {
    this.config = config;
  }

  /**
   * Variables #message and #number will be replaced in url as well as in parameter values.
   *
   * @return
   */
  public HttpResponseCode send(String phoneNumber, String message) {
    if (message == null) {
      log.error("Failed to send message to destination number: '" + StringHelper.hideStringEnding(phoneNumber, 'x', 3)
              + ". Message is null!");
      return HttpResponseCode.MESSAGE_ERROR;
    }
    if (message.length() > config.getSmsMaxMessageLength()) {
      log.error("Failed to send message to destination number: '" + StringHelper.hideStringEnding(phoneNumber, 'x', 3)
              + ". Message is to large, max length is " + config.getSmsMaxMessageLength() + ", but current message size is " + message.length());
      return HttpResponseCode.MESSAGE_TO_LARGE;
    }
    String proceededUrl = replaceVariables(config.getUrl(), phoneNumber, message, true);
    HttpMethodBase method = createHttpMethod(proceededUrl);
    if (config.getHttpMethodType() == SMSSenderConfig.HttpMethodType.GET) {
      if (MapUtils.isNotEmpty(config.getHttpParams())) {
        // Now build the query params list from the configured httpParams:
        NameValuePair[] params = new NameValuePair[config.getHttpParams().size()];
        int index = 0;
        for (Map.Entry<String, String> entry : config.getHttpParams().entrySet()) {
          String value = replaceVariables(entry.getValue(), phoneNumber, message, true);
          params[index++] = new NameValuePair(entry.getKey(), value);
        }
        ((GetMethod) method).setQueryString(params);
      }
    } else { // HTTP POST
      if (MapUtils.isNotEmpty(config.getHttpParams())) {
        // Now add all post params from the configured httpParams:
        for (Map.Entry<String, String> entry : config.getHttpParams().entrySet()) {
          String value = replaceVariables(entry.getValue(), phoneNumber, message, false);
          ((PostMethod) method).addParameter(entry.getKey(), value);
        }
      }
    }
    final HttpClient client = createHttpClient();
    try {
      int responseNumber = client.executeMethod(method);
      final String response = method.getResponseBodyAsString();
      log.info("Tried to send message to destination number: '" + StringHelper.hideStringEnding(phoneNumber, 'x', 3)
              + ". Response from service: " + response);
      HttpResponseCode responseCode = null;
      if (response == null || responseNumber != 200) {
        responseCode = HttpResponseCode.UNKNOWN_ERROR;
      } else if (matches(response, config.getSmsReturnPatternNumberError()) == true) {
        responseCode = HttpResponseCode.NUMBER_ERROR;
      } else if (matches(response, config.getSmsReturnPatternMessageToLargeError()) == true) {
        responseCode = HttpResponseCode.MESSAGE_TO_LARGE;
      } else if (matches(response, config.getSmsReturnPatternMessageError()) == true) {
        responseCode = HttpResponseCode.MESSAGE_ERROR;
      } else if (matches(response, config.getSmsReturnPatternError()) == true) {
        responseCode = HttpResponseCode.UNKNOWN_ERROR;
      } else if (matches(response, config.getSmsReturnPatternSuccess()) == true) {
        responseCode = HttpResponseCode.SUCCESS;
      } else {
        responseCode = HttpResponseCode.UNKNOWN_ERROR;
      }
      if (responseCode != HttpResponseCode.SUCCESS) {
        log.error("Unexpected response from sms gateway: " + responseNumber + ": " + response + " (if this call was successful, did you configured projectforge.sms.returnCodePattern.success?).");
      }
      return responseCode;
    } catch (final HttpException ex) {
      String errorKey = "Call failed. Please contact administrator.";
      log.error(errorKey + ": " + proceededUrl
              + StringHelper.hideStringEnding(String.valueOf(phoneNumber), 'x', 3));
      throw new RuntimeException(ex);
    } catch (final IOException ex) {
      String errorKey = "Call failed. Please contact administrator.";
      log.error(errorKey + ": " + proceededUrl
              + StringHelper.hideStringEnding(String.valueOf(phoneNumber), 'x', 3));
      throw new RuntimeException(ex);
    } finally {
      method.releaseConnection();
    }
  }

  private boolean matches(String str, String regexp) {
    if (regexp == null || str == null) {
      return false;
    }
    return str.matches(regexp);
  }

  /**
   * Variables #number and #message will be replaced by the user's form input.
   *
   * @param str    The string to proceed.
   * @param number The extracted phone number (already preprocessed...)
   * @return The given str with replaced vars (if exists).
   */
  private String replaceVariables(String str, String number, String message, boolean urlEncode) {
    if (number == null) return "";
    str = StringUtils.replaceOnce(str, "#number", urlEncode ? encode(number) : number);
    str = StringUtils.replaceOnce(str, "#message", urlEncode ? encode(message) : message);
    return str;
  }

  /**
   * Uses UTF-8
   *
   * @param str
   * @see URLEncoder#encode(String, String)
   */
  static String encode(final String str) {
    if (str == null) {
      return "";
    }
    try {
      return URLEncoder.encode(str, "UTF-8");
    } catch (final UnsupportedEncodingException ex) {
      log.info("Can't URL-encode '" + str + "': " + ex.getMessage());
      return "";
    }
  }

  /**
   * Used also for mocking {@link GetMethod} and {@link PostMethod}.
   *
   * @param url
   * @return
   */
  protected HttpMethodBase createHttpMethod(String url) {
    if (config.getHttpMethodType() == SMSSenderConfig.HttpMethodType.GET) {
      return new GetMethod(url);
    }
    return new PostMethod(url);
  }

  protected HttpClient createHttpClient() {
    return new HttpClient();
  }

  public SMSSender setConfig(SMSSenderConfig config) {
    this.config = config;
    return this;
  }
}

File: projectforge-business/src/main/kotlin/org/projectforge/messaging/SmsSender.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2023 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.messaging

import mu.KotlinLogging
import org.apache.commons.collections4.MapUtils
import org.apache.commons.io.IOUtils
import org.apache.commons.lang3.StringUtils
import org.apache.http.Consts
import org.apache.http.HttpStatus
import org.apache.http.NameValuePair
import org.apache.http.client.entity.UrlEncodedFormEntity
import org.apache.http.client.methods.HttpGet
import org.apache.http.client.methods.HttpPost
import org.apache.http.client.methods.HttpRequestBase
import org.apache.http.client.utils.URIBuilder
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.impl.client.HttpClients
import org.apache.http.message.BasicNameValuePair
import org.projectforge.common.StringHelper
import org.projectforge.sms.SmsSenderConfig
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.net.URISyntaxException
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

private val log = KotlinLogging.logger {}

class SmsSender(private var config: SmsSenderConfig) {
  enum class HttpResponseCode {
    SUCCESS, NUMBER_ERROR, MESSAGE_TO_LARGE, MESSAGE_ERROR, UNKNOWN_ERROR
  }

  /**
   * Variables #message and #number will be replaced in url as well as in parameter values.
   *
   * @return
   */
  fun send(phoneNumber: String?, message: String?): HttpResponseCode {
    if (message == null || phoneNumber == null) {
      log.error(
        "Failed to send message to destination number: '" + StringHelper.hideStringEnding(phoneNumber, 'x', 3)
            + ". Message is null!"
      )
      return HttpResponseCode.MESSAGE_ERROR
    }
    if (message.length > config.smsMaxMessageLength) {
      log.error(
        "Failed to send message to destination number: '" + StringHelper.hideStringEnding(phoneNumber, 'x', 3)
            + ". Message is to large, max length is " + config.smsMaxMessageLength + ", but current message size is " + message.length
      )
      return HttpResponseCode.MESSAGE_TO_LARGE
    }
    val proceededUrl = replaceVariables(config.url, phoneNumber, message, true)
    val method = createHttpMethod(proceededUrl, phoneNumber, message)
    createHttpClient().use { client ->
      return try {
        client.execute(method).use { httpResponse ->
          val statusCode = httpResponse.statusLine.statusCode
          var response: String? = null
          if (statusCode == HttpStatus.SC_OK) {
            val stream = httpResponse.entity.content
            response = IOUtils.toString(stream, StandardCharsets.UTF_8)
          }
          log.info(
            "Tried to send message to destination number: '" + StringHelper.hideStringEnding(phoneNumber, 'x', 3)
                + ". Response from service: " + response
          )
          val responseCode = if (response == null) {
            HttpResponseCode.UNKNOWN_ERROR
          } else if (matches(response, config.smsReturnPatternNumberError)) {
            HttpResponseCode.NUMBER_ERROR
          } else if (matches(response, config.smsReturnPatternMessageToLargeError)) {
            HttpResponseCode.MESSAGE_TO_LARGE
          } else if (matches(response, config.smsReturnPatternMessageError)) {
            HttpResponseCode.MESSAGE_ERROR
          } else if (matches(response, config.smsReturnPatternError)) {
            HttpResponseCode.UNKNOWN_ERROR
          } else if (matches(response, config.smsReturnPatternSuccess)) {
            HttpResponseCode.SUCCESS
          } else {
            HttpResponseCode.UNKNOWN_ERROR
          }
          if (responseCode != HttpResponseCode.SUCCESS) {
            log.error("Unexpected response from sms gateway: $statusCode: $response (if this call was successful, did you configured projectforge.sms.returnCodePattern.success?).")
          }
          responseCode
        }
      } catch (ex: IOException) {
        val errorKey = "Call failed. Please contact administrator."
        log.error(
          errorKey + ": " + proceededUrl + " for number "
              + StringHelper.hideStringEnding(phoneNumber, 'x', 3)
        )
        HttpResponseCode.UNKNOWN_ERROR
      }
    }
  }

  fun getErrorMessage(responseCode: HttpResponseCode?): String? {
    return if (responseCode == null) {
      "address.sendSms.sendMessage.result.unknownError"
    } else if (responseCode == HttpResponseCode.SUCCESS) {
      null
    } else if (responseCode == HttpResponseCode.MESSAGE_ERROR) {
      "address.sendSms.sendMessage.result.messageError"
    } else if (responseCode == HttpResponseCode.NUMBER_ERROR) {
      "address.sendSms.sendMessage.result.wrongOrMissingNumber"
    } else if (responseCode == HttpResponseCode.MESSAGE_TO_LARGE) {
      "address.sendSms.sendMessage.result.messageToLarge"
    } else {
      "address.sendSms.sendMessage.result.unknownError"
    }
  }

  private fun matches(str: String?, regexp: String?): Boolean {
    return if (regexp == null || str == null) {
      false
    } else str.matches(regexp.toRegex())
  }

  /**
   * Variables #number and #message will be replaced by the user's form input.
   *
   * @param str    The string to proceed.
   * @param number The extracted phone number (already preprocessed...)
   * @return The given str with replaced vars (if exists).
   */
  private fun replaceVariables(str: String?, number: String?, message: String, urlEncode: Boolean): String? {
    var str = str
    if (number == null) return ""
    str = StringUtils.replaceOnce(str, "#number", if (urlEncode) encode(number) else number)
    str = StringUtils.replaceOnce(str, "#message", if (urlEncode) encode(message) else message)
    return str
  }

  /**
   * Used also for mocking [HttpGet] and [HttpPost].
   *
   * @param url
   * @return
   */
  protected fun createHttpMethod(url: String?, phoneNumber: String?, message: String): HttpRequestBase {
    if (config.httpMethodType === SmsSenderConfig.HttpMethodType.GET) {
      return try {
        val uriBuilder = URIBuilder(url)
        if (MapUtils.isNotEmpty(config.httpParams)) {
          // Now build the query params list from the configured httpParams:
          val params = arrayOfNulls<NameValuePair>(
            config.httpParams!!.size
          )
          for ((key, value1) in config.httpParams!!) {
            val value = replaceVariables(value1, phoneNumber, message, true)
            uriBuilder.setParameter(key, value)
          }
        }
        HttpGet(uriBuilder.build())
      } catch (ex: URISyntaxException) {
        log.error("Configuration error, can't build url: " + ex.message, ex)
        throw RuntimeException(ex)
      }
    }
    // HTTP POST
    val post = HttpPost(url)
    if (MapUtils.isNotEmpty(config.httpParams)) {
      // Now add all post params from the configured httpParams:
      val formparams: MutableList<NameValuePair> = ArrayList()
      for ((key, value1) in config.httpParams!!) {
        val value = replaceVariables(value1, phoneNumber, message, false)
        formparams.add(BasicNameValuePair(key, value))
        val entity = UrlEncodedFormEntity(formparams, Consts.UTF_8)
        post.entity = entity
      }
    }
    return post
  }

  protected fun createHttpClient(): CloseableHttpClient {
    return HttpClients.createDefault()
  }

  fun setConfig(config: SmsSenderConfig): SmsSender {
    this.config = config
    return this
  }

  companion object {
    /**
     * Uses UTF-8
     *
     * @param str
     * @see URLEncoder.encode
     */
    fun encode(str: String?): String {
      return if (str == null) {
        ""
      } else try {
        URLEncoder.encode(str, "UTF-8")
      } catch (ex: UnsupportedEncodingException) {
        log.info("Can't URL-encode '" + str + "': " + ex.message)
        ""
      }
    }
  }
}
